src/replay/service.zig

```diff
@@ -38,162 +37,435 @@ const MAX_ENTRIES: u64 = 1024 * 1024; // 1 million slots is about 5 days
+        return .{
+            .allocator = self.allocator,
+            .logger = .from(self.logger),
+            .my_identity = self.my_identity,
+            .vote_account = null, // voting not currently supported
+
+            .account_store = self.account_store,
+            .thread_pool = &self.thread_pool,
+            .blockstore_reader = self.ledger.reader,
+            .slot_tracker = &self.slot_tracker,
+            .epochs = &self.epochs,
+            .progress_map = &self.progress,
+            .status_cache = &self.status_cache,
+        };
     }
 };
 
+/// Analogous to [`initialize_progress_and_fork_choice_with_locked_bank_forks`](https://github.com/anza-xyz/agave/blob/0315eb6adc87229654159448344972cbe484d0c7/core/src/replay_stage.rs#L637)
+fn initProgressAndForkChoiceWithLockedSlotForks(
+    allocator: std.mem.Allocator,
+    logger: Logger,
+    slot_tracker: *const SlotTracker,
+    epoch_tracker: *const EpochTracker,
+    my_pubkey: Pubkey,
+    vote_account: Pubkey,
+    blockstore: BlockstoreReader,
+) !struct { ProgressMap, HeaviestSubtreeForkChoice } {
+    const root_slot, const root_hash = blk: {
+        const root = slot_tracker.getRoot();
+        const root_slot = slot_tracker.root;
+        const root_hash = root.state.hash.readCopy();
+        break :blk .{ root_slot, root_hash.? };
+    };
+
+    var frozen_slots = try slot_tracker.frozenSlots(allocator);
+    defer frozen_slots.deinit(allocator);
+    const FrozenSlotsSortCtx = struct {
+        slots: []const Slot,
+        pub fn lessThan(ctx: @This(), a_index: usize, b_index: usize) bool {
+            return ctx.slots[a_index] < ctx.slots[b_index];
+        }
+    };
+    frozen_slots.sort(FrozenSlotsSortCtx{ .slots = frozen_slots.keys() });
+
+    var progress: ProgressMap = .INIT;
+    errdefer progress.deinit(allocator);
+
+    // Initialize progress map with any root slots
+    for (frozen_slots.keys(), frozen_slots.values()) |slot, ref| {
+        const prev_leader_slot = progress.getSlotPrevLeaderSlot(ref.constants.parent_slot);
+        try progress.map.ensureUnusedCapacity(allocator, 1);
+        progress.map.putAssumeCapacity(slot, try .initFromInfo(allocator, .{
+            .slot_info = ref,
+            .epoch_stakes = &epoch_tracker.getPtrForSlot(slot).?.stakes,
+            .now = .now(),
+            .validator_identity = &my_pubkey,
+            .validator_vote_pubkey = &vote_account,
+            .prev_leader_slot = prev_leader_slot,
+            .num_blocks_on_fork = 0,
+            .num_dropped_blocks_on_fork = 0,
+        }));
+    }
+
+    // Given a root and a list of `frozen_slots` sorted smallest to greatest by slot,
+    // initialize a new HeaviestSubtreeForkChoice
+    //
+    // Analogous to [`new_from_frozen_banks`](https://github.com/anza-xyz/agave/blob/0315eb6adc87229654159448344972cbe484d0c7/core/src/consensus/heaviest_subtree_fork_choice.rs#L235)
+    var heaviest_subtree_fork_choice = fork_choice: {
```

Comment:
Might be an idea to move this to a function on 

I can see this functionality being used in two places in Agave

https://github.com/anza-xyz/agave/blob/0d78d8f4814ff7f8d4b9152beed2839a397991c7/core/src/consensus.rs#L370

and here

https://github.com/anza-xyz/agave/blob/0d78d8f4814ff7f8d4b9152beed2839a397991c7/core/src/consensus/heaviest_subtree_fork_choice.rs#L256-L270

Can be a follow up pr.